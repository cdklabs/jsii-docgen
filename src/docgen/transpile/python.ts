import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const Case = require('case');

// Helpers
const toSnakeCase = (text?: string) => {
  return Case.snake(text ?? '');
};

const formatImport = (type: transpile.TranspiledType) => {
  if (type.submodule) {
    return `from ${type.module} import ${type.submodule}`;
  }
  return `import ${type.module}`;
};

const formatArguments = (inputs: string[]) => {
  return inputs.length === 0 ? '()' : [
    '(',
    inputs.map(i => `  ${i}`).join(',\n'),
    ')',
  ].join('\n');
};

const formatInvocation = (
  type: transpile.TranspiledType,
  inputs: string[],
  method?: string,
) => {
  let target;
  if (type.submodule) {
    if (!type.namespace) {
      throw new Error(
        `Invalid type: ${type.fqn}: Types defined in a submodule (${type.submodule}) must have a namespace. `,
      );
    }
    // we don't include the submodule name here since it is
    // included in the namespace. this works because we import the submodule
    // in this case.
    // TODO - merge `formatInvocation` with `formatImport` since they are inherently coupled.
    target = `${type.namespace}.${type.name}`;
  } else {
    target = type.fqn;
  }

  if (method) {
    target = `${target}.${method}`;
  }
  return `${target}${formatArguments(inputs)}`;
};

const formatSignature = (name: string, inputs: string[]) => {
  const def = 'def ';
  return `${def}${name}${formatArguments(inputs)}`;
};

/**
 * Hack to convert a jsii property to a parameter for
 * python specific parameter expansion.
 */
const propertyToParameter = (
  callable: reflect.Callable,
  property: reflect.Property,
): reflect.Parameter => {
  return {
    docs: property.docs,
    method: callable,
    name: property.name,
    optional: property.optional,
    parentType: property.parentType,
    spec: property.spec,
    system: property.system,
    type: property.type,
    variadic: false,
  };
};

/**
 * A python transpiler.
 */
export class PythonTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.PYTHON);
  }

  public readme(readme: string): string {
    return readme;
  }

  public unionOf(types: string[]): string {
    return `${this.typing('Union')}[${types.join(', ')}]`;
  }

  public listOf(type: string): string {
    return `${this.typing('List')}[${type}]`;
  }

  public mapOf(type: string): string {
    return `${this.typing('Mapping')}[${type}]`;
  }

  public any(): string {
    return this.typing('Any');
  }

  public boolean(): string {
    return 'bool';
  }

  public str(): string {
    return 'str';
  }

  public number(): string {
    return `${this.typing('Union')}[int, float]`;
  }

  public date(): string {
    return 'datetime.datetime';
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }
  public json(): string {
    return 'any';
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    const name = property.const ? property.name : toSnakeCase(property.name);
    const typeRef = this.typeReference(property.type);
    return {
      name,
      parentType: this.type(property.parentType),
      typeReference: typeRef,
      optional: property.optional,
      declaration: this.formatProperty(name, typeRef),
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    const name = toSnakeCase(parameter.name);
    const typeRef = this.typeReference(parameter.type);
    return {
      name,
      parentType: this.type(parameter.parentType),
      typeReference: typeRef,
      optional: parameter.optional,
      declaration: this.formatProperty(name, typeRef),
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const inputs = struct.allProperties.map((p) =>
      this.formatParameters(this.property(p)),
    );
    return {
      type: type,
      name: struct.name,
      import: formatImport(type),
      initialization: formatInvocation(type, inputs),
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);

    const parameters = new Array<reflect.Parameter>();

    for (const p of callable.parameters.sort(this.optionalityCompare)) {
      if (!this.isStruct(p)) {
        parameters.push(p);
      } else {
        // struct parameters are expanded to the individual struct properties
        const struct = p.parentType.system.findInterface(p.type.fqn!);
        for (const property of struct.allProperties) {
          const parameter = propertyToParameter(callable, property);
          parameters.push(parameter);
        }
      }
    }

    const name = toSnakeCase(callable.name);
    const inputs = parameters.map((p) => this.formatParameters(this.parameter(p)));

    return {
      name,
      parentType: type,
      import: formatImport(type),
      parameters,
      signatures: [formatSignature(name, inputs)],
      invocations: [formatInvocation(
        type,
        inputs,
        callable.kind === reflect.MemberKind.Initializer ? undefined : name,
      )],
    };
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    const fqn = moduleLike.targets?.python?.module;
    if (!fqn) {
      throw new Error(
        `Python is not a supported target for module: ${moduleLike.fqn}`,
      );
    }

    if (moduleLike instanceof reflect.Submodule) {
      const fqnParts = fqn.split('.');
      return { name: fqnParts[0], submodule: fqnParts[1] };
    }
    return { name: fqn };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter,
  ): number {
    if (!p1.optional && p2.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private isStruct(p: reflect.Parameter): boolean {
    return p.type.fqn ? p.system.findFqn(p.type.fqn).isDataType() : false;
  }

  private typing(type: 'List' | 'Mapping' | 'Any' | 'Union'): string {
    return `typing.${type}`;
  }

  private formatParameters(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${transpiled.name}: ${tf}${transpiled.optional ? ' = None' : ''}`;
  }

  private formatProperty(
    name: string,
    typeReference: transpile.TranspiledTypeReference,
  ): string {
    const tf = typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${name}: ${tf}`;
  }
}
