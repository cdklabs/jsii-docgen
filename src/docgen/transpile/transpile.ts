import * as reflect from 'jsii-reflect';

/**
 * Supported languages to generate documentation in.
 */
export class Language {
  /**
   * TypeScript.
   */
  public static readonly TYPESCRIPT = new Language('typescript');

  /**
   * Python.
   */
  public static readonly PYTHON = new Language('python');

  /**
   * Transform a literal string to the `Language` object.
   *
   * Throws an `UnsupportedLanguageError` if the language is not supported.
   */
  public static fromString(lang: string) {
    switch (lang) {
      case Language.TYPESCRIPT.toString():
        return Language.TYPESCRIPT;
      case Language.PYTHON.toString():
        return Language.PYTHON;
      default:
        throw new UnsupportedLanguageError(lang, [Language.TYPESCRIPT, Language.PYTHON]);
    }
  }

  private constructor(private readonly lang: string) {}

  public toString() {
    return this.lang;
  }
}

export class UnsupportedLanguageError extends Error {
  constructor(lang: string, supported: Language[]) {
    super(`Unsupported language: ${lang}. Supported languages are: [${supported}]`);
  }
}

/**
 * Outcome of transpiling a jsii struct.
 */
export interface TranspiledStruct {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The import statement needed in order to use this struct.
   */
  readonly import: string;
  /**
   * How to initialize this struct.
   */
  readonly initialization: string;
}

/**
 * Outcome of transpiling a jsii class.
 */
export interface TranspiledClass {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a jsii callable.
 */
export interface TranspiledCallable {
  /**
   * The (transpiled) parent type.
   */
  readonly parentType: TranspiledType;
  /**
   * How a signature of the callable looks like.
   */
  readonly signature: string;
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The import statement needed in order to use this callable.
   */
  readonly import: string;
  /**
   * How an invocation of this callable looks like.
   */
  readonly invocation: string;
  /**
   * The jsii parameters this callable accepts.
   */
  readonly parameters: reflect.Parameter[];
}

/**
 * Outcome of transpiling a jsii parameter.
 */
export interface TranspiledParameter {
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The (transpiled) parent type.
   */
  readonly parentType: TranspiledType;
  /**
   * The (transpiled) type reference.
   */
  readonly typeReference: TranspiledTypeReference;
  /**
   * Whether or not the parameter is optional.
   */
  readonly optional: boolean;
}

/**
 * Outcome of transpiling a jsii interface (not data type).
 */
export interface TranspiledInterface {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a generic jsii type.
 */
export interface TranspiledType {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * Simple name of the type.
   */
  readonly name: string;
  /**
   * Namespace of the type.
   */
  readonly namespace?: string;
  /**
   * The language specific module name the type belongs to.
   */
  readonly module: string;
  /**
   * The language specific submodule name the type belongs to.
   */
  readonly submodule?: string;
}

/**
 * Options for how to render a string representation of a type reference.
 */
export interface TranspiledTypeReferenceToStringOptions {
  /**
   * Type formatter.
   */
  typeFormatter?: (type: TranspiledType) => string;
  /**
   * String formatter.
   */
  stringFormatter?: (typeName: string) => string;
}

/**
 * Outcome of transpiling a jsii type reference.
 */
export class TranspiledTypeReference {
  /**
   * Create a type reference that reprensents a primitive.
   */
  public static primitive(
    transpile: Transpile,
    ref: reflect.TypeReference,
    primitive: string,
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(transpile, ref, primitive);
  }
  /**
   * Create a type reference that represents any type.
   */
  public static any(
    transpile: Transpile,
    ref: reflect.TypeReference,
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(transpile, ref, undefined, true);
  }
  /**
   * Create a type reference that reprenets a concrete type.
   */
  public static type(
    transpile: Transpile,
    ref: reflect.TypeReference,
    type: TranspiledType,
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      type,
    );
  }
  /**
   * Create a type reference that reprenets an array of a type reference.
   */
  public static arrayOfType(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tf: TranspiledTypeReference,
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      tf,
    );
  }
  /**
   * Create a type reference that reprenets a map of a type reference.
   */
  public static mapOfType(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tf: TranspiledTypeReference,
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      undefined,
      tf,
    );
  }
  /**
   * Create a type reference that reprenets a union of a type references.
   */
  public static unionOfTypes(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tfs: TranspiledTypeReference[],
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      tfs,
    );
  }

  private constructor(
    /**
     * A transpiler
     */
    private readonly transpile: Transpile,
    /**
     * The original type reference.
     */
    private readonly ref: reflect.TypeReference,
    /**
     * Primitive type ref.
     */
    private readonly primitive?: string,
    /**
     * 'Any' type ref
     */
    private readonly isAny?: boolean,
    /**
     * Concrete type.
     */
    private readonly type?: TranspiledType,
    /**
     * Array of ref.
     */
    private readonly arrayOfType?: TranspiledTypeReference,
    /**
     * Map of ref.
     */
    private readonly mapOfType?: TranspiledTypeReference,
    /**
     * Union of ref.
     */
    private readonly unionOfTypes?: TranspiledTypeReference[],
  ) {}

  public toString(options?: TranspiledTypeReferenceToStringOptions): string {
    const tFormatter = options?.typeFormatter ?? ((t) => t.fqn);
    const sFormatter = options?.stringFormatter ?? ((s) => s);
    if (this.primitive) {
      return sFormatter(this.primitive);
    }
    if (this.type) {
      return tFormatter(this.type);
    }
    if (this.isAny) {
      return sFormatter(this.transpile.any());
    }
    if (this.arrayOfType) {
      const ref = this.arrayOfType.toString(options);
      return this.transpile.listOf(ref);
    }
    if (this.mapOfType) {
      const ref = this.mapOfType.toString(options);
      return this.transpile.mapOf(ref);
    }
    if (this.unionOfTypes) {
      const refs = this.unionOfTypes.map((t) => t.toString(options));
      return this.transpile.unionOf(refs);
    }
    throw new Error(`Invalid type reference: ${this.ref.toString()}`);
  }
}

/**
 * Outcome of transpiling a jsii property.
 */
export type TranspiledProperty = TranspiledParameter;

/**
 * Outcome of transpiling a jsii enum.
 */
export interface TranspiledEnum {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a specific enum member.
 */
export interface TranspiledEnumMember {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a module like object. (Assembly | Submodule)
 */
export interface TranspiledModuleLike {
  /**
   * The language specific module name.
   *
   * In case the module like object is a submodule, this should contain
   * only the root module name.
   *
   * In case the module like object is the root module, this should contain
   * the module fqn.
   *
   * Examples:
   *
   *   `aws-cdk-lib` -> `aws_cdk`
   *   `aws-cdk-lib.aws_eks` -> `aws_cdk`
   *   `@aws-cdk/aws-eks` -> `aws_cdk.aws_eks`
   */
  readonly name: string;
  /**
   * The language specific submodule name.
   *
   * In case the module like object is a submodule, this should contain
   * only the submodule name.
   *
   * In case the module like object is the root module, this should be undefined
   *
   * Examples:
   *
   *   `aws-cdk-lib` -> undefined
   *   `aws-cdk-lib.aws_eks` -> `aws_eks`
   *   `@aws-cdk/aws-eks` -> undefined
   */
  readonly submodule?: string;
}

/**
 * Language transpiling for jsii types.
 */
export interface Transpile {
  /**
   * The language of the transpiler.
   */
  readonly language: Language;

  /**
   * How links to types should be formatted.
   *
   * @default '#{fqn}'
   */
  readonly linkFormatter?: (type: TranspiledType) => string;

  /**
   * Transpile a module like object (Assembly | Submodule)
   */
  moduleLike(moduleLike: reflect.ModuleLike): TranspiledModuleLike;

  /**
   * Transpile a callable (method, static function, initializer)
   */
  callable(callable: reflect.Callable): TranspiledCallable;

  /**
   * Transpile a class.
   */
  class(klass: reflect.ClassType): TranspiledClass;

  /**
   * Transpile a struct (data type interface).
   */
  struct(struct: reflect.InterfaceType): TranspiledStruct;

  /**
   * Transpile an interface (non data type).
   */
  interface(iface: reflect.InterfaceType): TranspiledInterface;

  /**
   * Transpile a parameter.
   */
  parameter(parameter: reflect.Parameter): TranspiledParameter;

  /**
   * Transpile a property.
   */
  property(property: reflect.Property): TranspiledProperty;

  /**
   * Transpile an enum.
   */
  enum(enu: reflect.EnumType): TranspiledEnum;

  /**
   * Transpile an enum member.
   */
  enumMember(em: reflect.EnumMember): TranspiledEnumMember;

  /**
   * Transpile a type.
   */
  type(type: reflect.Type): TranspiledType;

  /**
   * Transpile (recursively) a type reference.
   */
  typeReference(typeReference: reflect.TypeReference): TranspiledTypeReference;

  /**
   * How a union looks like in the target language.
   */
  unionOf(types: string[]): string;

  /**
   * How a list looks like in the target language.
   */
  listOf(type: string): string;

  /**
   * How a map looks like in the target language.
   */
  mapOf(type: string): string;

  /**
   * How the 'any' type is represented in the target language.
   */
  any(): string;

  /**
   * How the 'string' type is represented in the target language.
   */
  str(): string;

  /**
   * How the 'number' type is represented in the target language.
   */
  number(): string;

  /**
   * How the 'boolean' type is represented in the target language.
   */
  boolean(): string;

  /**
   * How the 'json' type is represented in the target language.
   */
  json(): string;

  /**
   * How the 'date' type is represented in the target language.
   */
  date(): string;

  /**
   * How a readme is displayed in the target language.
   */
  readme(readme: string): string;
}

// interface merging so we don't have to implement these methods
// in the abstract class.
export interface TranspileBase extends Transpile {}

/**
 * Common functionality between different transpilers.
 */
export abstract class TranspileBase implements Transpile {
  constructor(public readonly language: Language) {}

  public type(type: reflect.Type): TranspiledType {
    const submodule = this.findSubmodule(type);
    const moduleLike = this.moduleLike(submodule ? submodule : type.assembly);

    const fqn = [moduleLike.name];

    // TODO - these should also probably be transliterated.
    // for python it just happens to work
    if (type.namespace) {
      fqn.push(type.namespace);
    }
    fqn.push(type.name);

    return {
      fqn: fqn.join('.'),
      name: type.name,
      namespace: type.namespace,
      module: moduleLike.name,
      submodule: moduleLike.submodule,
    };
  }

  public typeReference(ref: reflect.TypeReference): TranspiledTypeReference {
    if (ref.type) {
      const transpiled = this.type(ref.type);
      return TranspiledTypeReference.type(this, ref, transpiled);
    }

    if (ref.unionOfTypes) {
      const transpiled = ref.unionOfTypes.map((t) => this.typeReference(t));
      return TranspiledTypeReference.unionOfTypes(this, ref, transpiled);
    }

    if (ref.arrayOfType) {
      const transpiled = this.typeReference(ref.arrayOfType);
      return TranspiledTypeReference.arrayOfType(this, ref, transpiled);
    }

    if (ref.mapOfType) {
      const transpiled = this.typeReference(ref.mapOfType);
      return TranspiledTypeReference.mapOfType(this, ref, transpiled);
    }

    if (ref.isAny) {
      return TranspiledTypeReference.any(this, ref);
    }

    if (ref.primitive) {
      let transpiled;
      switch (ref.primitive) {
        case 'string':
          transpiled = this.str();
          break;
        case 'boolean':
          transpiled = this.boolean();
          break;
        case 'number':
          transpiled = this.number();
          break;
        case 'date':
          transpiled = this.date();
          break;
        case 'json':
          transpiled = this.json();
          break;
        default:
          throw new Error(`Unsupported primitive type '${ref.primitive}'`);
      }
      return TranspiledTypeReference.primitive(this, ref, transpiled);
    }

    throw new Error(`Unsupported type: ${ref.toString()}`);
  }

  protected findSubmodule(type: reflect.Type): reflect.Submodule | undefined {
    if (!type.namespace) {
      return undefined;
    }

    // if the type is in a submodule, the submodule name is the first
    // part of the namespace. we construct the full submodule fqn and search for it.
    const submoduleFqn = `${type.assembly.name}.${
      type.namespace.split('.')[0]
    }`;
    const submodules = type.assembly.submodules.filter(
      (s) => s.fqn === submoduleFqn,
    );

    if (submodules.length > 1) {
      // can never happen, but the array data structure forces this handling.
      throw new Error(`Found multiple submodulues with fqn ${submoduleFqn}`);
    }

    if (submodules.length === 0) {
      return undefined;
    }

    // type is inside this submodule.
    return submodules[0];
  }
}
