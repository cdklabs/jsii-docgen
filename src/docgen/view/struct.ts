import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { Transpile, TranspiledStruct, TranspiledType } from '../transpile/transpile';
import { Property } from './property';

export class Struct {
  private readonly transpiled: TranspiledStruct;
  private readonly properties: Property[] = new Array<Property>();
  constructor(
    private readonly transpile: Transpile,
    private readonly iface: reflect.InterfaceType,
    linkFormatter: (type: TranspiledType) => string,
  ) {
    this.transpiled = transpile.struct(iface);
    for (const property of this.iface.allProperties) {
      this.properties.push(new Property(this.transpile, property, linkFormatter));
    }
  }

  public render(): Markdown {
    const md = new Markdown({
      id: this.transpiled.type.fqn,
      header: { title: this.transpiled.name },
    });

    if (this.iface.docs) {
      md.docs(this.iface.docs);
    }

    const initializer = new Markdown({
      id: `${this.transpiled.type}.Initializer`,
      header: { title: 'Initializer' },
    });

    initializer.code(
      this.transpile.language.toString(),
      `${this.transpiled.import}`,
      '',
      `${this.transpiled.initialization}`,
    );

    for (const property of this.properties) {
      initializer.section(property.render());
    }

    md.section(initializer);
    return md;
  }
}
