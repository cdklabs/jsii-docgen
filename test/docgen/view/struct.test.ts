import * as reflect from 'jsii-reflect';
import { PythonTranspile } from '../../../src/docgen/transpile/python';
import { TranspiledType } from '../../../src/docgen/transpile/transpile';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { Struct } from '../../../src/docgen/view/struct';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

describe('python', () => {
  const transpile = new PythonTranspile();
  test('snapshot', () => {
    const struct = new Struct(
      transpile,
      assembly.system.interfaces.filter((i) => i.datatype)[0],
      (t: TranspiledType) => `#${t.fqn}`,
    );
    expect(struct.render().render()).toMatchSnapshot();
  });
});

describe('typescript', () => {
  const transpile = new TypeScriptTranspile();
  test('snapshot', () => {
    const struct = new Struct(
      transpile,
      assembly.system.interfaces.filter((i) => i.datatype)[0],
      (t: TranspiledType) => `#${t.fqn}`,
    );
    expect(struct.render().render()).toMatchSnapshot();
  });
});
