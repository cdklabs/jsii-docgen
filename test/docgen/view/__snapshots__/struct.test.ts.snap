// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`java snapshot 1`] = `
" CfnPublicRepositoryProps <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps\\"></a>

Properties for defining a \`AWS::ECR::PublicRepository\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html)

# Initializer <a name=\\"[object Object].Initializer\\"></a>

\`\`\`java
import software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps;

CfnPublicRepositoryProps.builder()
//  .repositoryCatalogData(java.lang.Object)
//  .repositoryName(java.lang.String)
//  .repositoryPolicyText(java.lang.Object)
//  .tags(java.util.List<CfnTag>)
    .build();
\`\`\`

## \`repositoryCatalogData\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.property.repositoryCatalogData\\"></a>

- *Type:* \`java.lang.Object\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---

## \`repositoryName\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.property.repositoryName\\"></a>

- *Type:* \`java.lang.String\`

\`AWS::ECR::PublicRepository.RepositoryName\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname)

---

## \`repositoryPolicyText\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.property.repositoryPolicyText\\"></a>

- *Type:* \`java.lang.Object\`

\`AWS::ECR::PublicRepository.RepositoryPolicyText\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext)

---

## \`tags\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.property.tags\\"></a>

- *Type:* java.util.List<[\`software.amazon.awscdk.core.CfnTag\`](#software.amazon.awscdk.core.CfnTag)>

\`AWS::ECR::PublicRepository.Tags\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags)

---
"
`;

exports[`python snapshot 1`] = `
" CfnPublicRepositoryProps <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps\\"></a>

Properties for defining a \`AWS::ECR::PublicRepository\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html)

# Initializer <a name=\\"[object Object].Initializer\\"></a>

\`\`\`python
import aws_cdk.aws_ecr

aws_cdk.aws_ecr.CfnPublicRepositoryProps(
  repository_catalog_data: typing.Any = None,
  repository_name: str = None,
  repository_policy_text: typing.Any = None,
  tags: typing.List[CfnTag] = None
)
\`\`\`

## \`repository_catalog_data\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.property.repository_catalog_data\\"></a>

- *Type:* \`typing.Any\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---

## \`repository_name\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.property.repository_name\\"></a>

- *Type:* \`str\`

\`AWS::ECR::PublicRepository.RepositoryName\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname)

---

## \`repository_policy_text\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.property.repository_policy_text\\"></a>

- *Type:* \`typing.Any\`

\`AWS::ECR::PublicRepository.RepositoryPolicyText\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext)

---

## \`tags\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.property.tags\\"></a>

- *Type:* typing.List[[\`aws_cdk.core.CfnTag\`](#aws_cdk.core.CfnTag)]

\`AWS::ECR::PublicRepository.Tags\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags)

---
"
`;

exports[`typescript snapshot 1`] = `
" CfnPublicRepositoryProps <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps\\"></a>

Properties for defining a \`AWS::ECR::PublicRepository\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html)

# Initializer <a name=\\"[object Object].Initializer\\"></a>

\`\`\`typescript
import { CfnPublicRepositoryProps } from '@aws-cdk/aws-ecr'

const cfnPublicRepositoryProps: CfnPublicRepositoryProps = { ... }
\`\`\`

## \`repositoryCatalogData\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps.property.repositoryCatalogData\\"></a>

- *Type:* \`any\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---

## \`repositoryName\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps.property.repositoryName\\"></a>

- *Type:* \`string\`

\`AWS::ECR::PublicRepository.RepositoryName\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname)

---

## \`repositoryPolicyText\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps.property.repositoryPolicyText\\"></a>

- *Type:* \`any\`

\`AWS::ECR::PublicRepository.RepositoryPolicyText\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext)

---

## \`tags\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepositoryProps.property.tags\\"></a>

- *Type:* [\`@aws-cdk/core.CfnTag\`](#@aws-cdk/core.CfnTag)[]

\`AWS::ECR::PublicRepository.Tags\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags)

---
"
`;
